File: src/pgo/model/tla/TLAQuantifierBound.java
package pgo.model.tla;

import java.util.List;
import java.util.stream.Collectors;

import pgo.util.SourceLocation;

public class TLAQuantifierBound extends TLANode {
	
	List<TLAIdentifier> ids;
	TLAExpression set;
	Type type;

	public TLAQuantifierBound(SourceLocation location, Type type, List<TLAIdentifier> ids, TLAExpression set) {
		super(location);
		this.type = type;
		this.ids = ids;
		this.set = set;
	}
	
	public enum Type{
		IDS,
		TUPLE,
	}
	
	@Override
	public TLAQuantifierBound copy() {
		return new TLAQuantifierBound(getLocation(), type, ids.stream().map(TLAIdentifier::copy).collect(Collectors.toList()), set.copy());
	}
	
	public Type getType() {
		return type;
	}
	
	public List<TLAIdentifier> getIds(){
		return ids;
	}
	
	public TLAExpression getSet() {
		return set;
	}
	
	@Override
	public <T, E extends Throwable> T accept(TLANodeVisitor<T, E> v) throws E {
		return v.visit(this);
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((ids == null) ? 0 : ids.hashCode());
		result = prime * result + ((set == null) ? 0 : set.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		TLAQuantifierBound other = (TLAQuantifierBound) obj;
		if (ids == null) {
			if (other.ids != null)
				return false;
		} else if (!ids.equals(other.ids))
			return false;
		if (set == null) {
			return other.set == null;
		} else return set.equals(other.set);
	}

}

File: src/pgo/model/tla/TLAQuantifierBound.scala
package pgo.model.tla

import scala.jdk.CollectionConverters._

import pgo.util.SourceLocation

final class TLAQuantifierBound(loc: SourceLocation, val tpe: TLAQuantifierBound.Type, val ids: List[TLAIdentifier], val set: TLAExpression) extends TLANode(loc) with TLADefinitionComposite {
  def getType: TLAQuantifierBound.Type = tpe
  def getIds: java.util.List[TLAIdentifier] = ids.asJava
  def getSet: TLAExpression = set

  lazy val identifierOrTuple: TLAIdentifierOrTuple =
    tpe match {
      case TLAQuantifierBound.Type.IDS =>
        assert(ids.length == 1)
        TLAIdentifierOrTuple.Identifier(ids.head)
      case TLAQuantifierBound.Type.TUPLE =>
        TLAIdentifierOrTuple.Tuple(loc, ids.asJava)
    }

  def this(loc: SourceLocation, tpe: TLAQuantifierBound.Type, ids: java.util.List[TLAIdentifier], set: TLAExpression) =
    this(loc, tpe, ids.asScala.toList, set)

  override def copy: TLAQuantifierBound = new TLAQuantifierBound(loc, tpe, ids.map(_.copy), set.copy())

  override def accept[T, E <: Throwable](v: TLANodeVisitor[T, E]): T = v.visit(this)

  override val members: List[TLADefinition] = ids

  override def equals(other: Any): Boolean = other match {
    case that: TLAQuantifierBound =>
      tpe == that.tpe &&
        ids == that.ids &&
        set == that.set
    case _ => false
  }

  override def hashCode(): Int = Seq(tpe, ids, set).hashCode()
}

object TLAQuantifierBound {
  sealed abstract class Type
  object Type {
    final case object IDS extends Type
    def ids: IDS.type = IDS
    final case object TUPLE extends Type
    def tuple: TUPLE.type = TUPLE
  }
}
